src/iter.rs

```diff
@@ -0,0 +1,254 @@
+
+use crate::index::{IndexType, NodeIndex};
+use crate::interval::Interval;
+use crate::intervalmap::IntervalMap;
+use crate::node::Node;
+
+/// Pushes a link of nodes on the left to stack.
+fn left_link<T, V, Ix>(map_ref: &IntervalMap<T, V, Ix>, mut x: NodeIndex<Ix>) -> Vec<NodeIndex<Ix>>
+where
+    T: Ord,
+    Ix: IndexType,
+{
+    let mut nodes = vec![];
+    while !map_ref.node_ref(x, Node::is_sentinel) {
+        nodes.push(x);
+        x = map_ref.node_ref(x, Node::left);
+    }
+    nodes
+}
+
+/// An iterator over the entries of a `IntervalMap`.
+#[derive(Debug)]
+pub struct Iter<'a, T, V, Ix>
+where
+    T: Ord,
+{
+    /// Reference to the map
+    pub(crate) map_ref: &'a IntervalMap<T, V, Ix>,
+    /// Stack for iteration
+    pub(crate) stack: Vec<NodeIndex<Ix>>,
+}
+
+impl<'a, T, V, Ix> Iter<'a, T, V, Ix>
+where
+    T: Ord,
+    Ix: IndexType,
+{
+    pub fn new(map_ref: &'a IntervalMap<T, V, Ix>) -> Self {
+        Iter {
+            map_ref,
+            stack: left_link(map_ref, map_ref.root),
+        }
+    }
+}
+
+impl<'a, T, V, Ix> Iterator for Iter<'a, T, V, Ix>
+where
+    T: Ord,
+    Ix: IndexType,
+{
+    type Item = (&'a Interval<T>, &'a V);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.stack.is_empty() {
+            return None;
+        }
+        let x = self.stack.pop().unwrap();
+        self.stack.extend(left_link(
+            self.map_ref,
+            self.map_ref.node_ref(x, Node::right),
+        ));
+        Some(self.map_ref.node_ref(x, |xn| (xn.interval(), xn.value())))
+    }
+}
+
+/// An into iterator over the entries of a `IntervalMap`.
+#[derive(Debug)]
+pub struct IntoIter<T, V, Ix>
+where
+    T: Ord,
+{
+    interval_map: IntervalMap<T, V, Ix>,
+    /// Stack for iteration
+    pub(crate) stack: Vec<NodeIndex<Ix>>,
+}
+
+impl<'a, T, V, Ix> IntoIter<T, V, Ix>
+where
+    T: Ord,
+    Ix: IndexType,
+{
+    pub fn new(interval_map: IntervalMap<T, V, Ix>) -> Self {
+        let mut temp = IntoIter {
+            interval_map,
+            stack: vec![],
+        };
+        temp.stack = left_link(&temp.interval_map, temp.interval_map.root);
+        temp
+    }
+}
+
+impl<'a, T, V, Ix> Iterator for IntoIter<T, V, Ix>
+where
+    T: Ord,
+    Ix: IndexType,
+{
+    type Item = (Interval<T>, V);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.stack.is_empty() {
+            return None;
+        }
+        let x = self.stack.pop().unwrap();
+        self.stack.extend(left_link(
+            &self.interval_map,
+            self.interval_map.node_ref(x, Node::right),
+        ));
+        let res = &mut self.interval_map.nodes[x.index()];
+        Some((
+            std::mem::replace(&mut res.interval.as_mut().unwrap(), unsafe {
+                std::mem::zeroed()
+            }),
+            std::mem::replace(&mut res.value.as_mut().unwrap(), unsafe {
+                std::mem::zeroed()
+            }),
+        ))
+    }
```

Comment:
Why must `unsafe` be used here?

Reply:
I want to move these value out and make them inaccessible. Interval and value don't have `Default` trait so I used `std::mem::zeroed()`, do you have any suggestions?

Reply:
use `Option::take`.

Reply:
> use `Option::take`.

What a fantastic suggestion! I spent a long time refining this ugly writing with gpt but failed:( Thanks a lot!